internal/solver/solver.go

```diff
@@ -0,0 +1,485 @@
+
+	// TODO
+	// Upgrade1ToMajor: same as Upgrade, but tune semver distances.
+	// Upgrade1ToMinor: same as Upgrade, but tune semver distances.
+	// UpgradeAll: don't add a constraintPresent for all the current releases,
+	//             with the AtMost1 of all the versions, plus semver distance, it's ok.
+	// UpgradeAllToMajor: same as UpgradeAll, but tune semver distances.
+	// UpgradeAllToMinor: same as UpgradeAll, but tune semver distances.
+	// Remove1: don't add a constraintPresent for the specific release/package to be removed.
+	//          CurrentStatus:Present and DesiredStatus:Absent
+	// CheckAll: add everything with desiredState:Unknown, check SAT or UNSAT.
+	// AutoremoveAll: all packages not marked as autoinstalled can be dropped. Not
+	//                enough with setting their desiredState to absent, they may be
+	//                dependencies.
+)
+
+type Solver struct {
+	PkgDB        *PkgDB       // DB containing packages
+	PkgResultSet PkgResultSet // outcome of sat solving
+	Strategy     SolverStrategy
+}
+
+// PkgResultSet contains the status outcome of solving, and the different sets of
+// packages derived from the outcome.
+// It will be marshalled into Yaml and Json.
+type PkgResultSet struct {
+	PresentUnchanged []*pkg.Pkg
+	ToInstall        []*pkg.Pkg
+	ToRemove         []*pkg.Pkg
+	Status           string
+	Inconsistencies  []string
+}
+
+type OutputMode int
+
+const (
+	JSON OutputMode = iota
+	YAML
+	Table
+)
+
+// New creates a new Solver, initializing its database.
+func New(strategy SolverStrategy) (s *Solver) {
+	s = &Solver{
+		PkgDB:        CreatePkgDBInstance(),
+		PkgResultSet: PkgResultSet{},
+		Strategy:     strategy,
+	}
+	s.PkgResultSet.Inconsistencies = []string{}
+	return s
+}
+
+// BuildWorldMock fills the database with pkgs instead of releases, charts from
+// repositories, and so.
+// Useful for testing.
+func (s *Solver) BuildWorldMock(pkgs []*pkg.Pkg) {
+	for _, p := range pkgs {
+		s.PkgDB.Add(p)
+	}
+}
+
+// BuildConstraints generates all constraints for package p
+func (s *Solver) BuildConstraints(p *pkg.Pkg) (constrs []maxsat.Constr) {
+
+	// add constraints for relationships
+	packageConstrs := s.buildConstraintRelations(p)
+	constrs = append(constrs, packageConstrs...)
+
+	if p.CurrentState == pkg.Present && p.DesiredState != pkg.Absent {
+		// p is a release, and is not going to be changed
+		packageConstrs := s.buildConstraintPresent(p)
+		constrs = append(constrs, packageConstrs...)
+	}
+
+	if p.DesiredState != pkg.Unknown {
+		// p is going to be installed, or removed (and is a release)
+		packageConstrs := s.buildConstraintToModify(p)
+		constrs = append(constrs, packageConstrs...)
+	}
+
+	// add constraints for all packages that are the same and differ only in version
+	packageConstrs = s.buildConstraintAtMost1(p)
+	constrs = append(constrs, packageConstrs...)
+
+	return constrs
+}
+
+func (s *Solver) Solve(logger log.Logger) {
```

Comment:
Why is the logger passed to this method? Would it be better to attach is to the `Solver` type and then log different method details, too?

Reply:
Doh, indeed. Fixed in https://github.com/rancher-sandbox/hypper/pull/176/commits/a6616a6578932de4e7890e40a45a4efae51efccd.

We should refactor more than that, but I would like to have all the tests passing beforehand.